docs_src/design/adr/device-service/0021-Profile-Changes.md

```diff
@@ -0,0 +1,126 @@
+
+Change of the device profile on a device – never allowed unless there are no events or readings in the system.  There is too much risk that an existing event or reading also contains the device profile reference – causing complexity in knowing what information is in the event/reading.
+
+A provision watcher also has an association to a device profile.  However, the device profile is only used to know the type of device to create in case of provisioning.  Therefore, if a device profile changes, it does not impact the provision watcher.
+
+A device profile cannot be deleted while it is associated to a provision watcher
+
+Change of the device profile on an event or reading is never allowed.  Events and readings are immutable.
+
+### Applicability
+
+This document dictates the rules around allowed change to device profiles (and associations).  All EdgeX services will abide by these rules (implementing appropriate checks and validation as necessary to enforce the rules) when this ADR is approved to include:
+- service APIs
+- user interfaces (like the Edge GUI)
+- client tools (like the EdgeX CLI)
+
+## Proposed Design
+
+The following metadata API changes are suggested as a means to implement the device profile rules of this ADR
+
+- Block Device DELETE when there is any associated event/reading exists.
+- Block Profile (Upload) PUT when there is any associated device/event/reading exists.
+- Block Profile DELETE when there is any associated device/event/reading exists.
+- Allow empty Profile POST (containing no device resources or commands)
+- Add Profile General Property PATCH API
+- Add Profile Device Resource POST API
+- Add Profile Device Resource PATCH API (allow to modify Description and IsHidden only)
+- Add Profile Device Resource DELETE API (check it's not associated to any device, reading or event)
+- Add Profile Device Command POST API
+- Add Profile Device Command PATCH API (allow to modify Description and IsHidden only)
+- Add Profile Device Command DELETE API (check it's not associated to any device, reading or event)
+
+*Note* - to be determined: how to allow for device resource or device command changes (outside of Description and isHidden) when the device profile is not yet associated to any object
+
+## Decision
+
+TBD
+
+## Consequences/Considerations
+
+The metadata device profile PUT and DELETE APIs must contain the validation logic per the terms above.
+
+- [Device Profile PUT](https://app.swaggerhub.com/apis/EdgeXFoundry1/core-metadata/2.0.0#/default/put_deviceprofile)
+- [Device Profile Upload PUT](https://app.swaggerhub.com/apis/EdgeXFoundry1/core-metadata/2.0.0#/default/put_deviceprofile_uploadfile)
+- [Device Profile DELETE](https://app.swaggerhub.com/apis/EdgeXFoundry1/core-metadata/2.0.0#/default/delete_deviceprofile_name__name_)
+
+The device PATCH API must validate that device profile is not changed unless there are no events or readings in core data
```

Comment:
Shouldn't this say "validate that the device is not changed"? I'm not sure what the device PATH API needs to check that it's profile isn't being changed?

Reply:
The thinking here is that if you have a device and it is already associated to events/readings (and therefore those events/readings are associated to the profile and its device resources) what would it mean to change the device's device profile?  You would then have events/readings that say they are associated to a profile and device resources that are not associated to that profile or device resources anymore.  It would be a referential integrity issue again.  

Unless we want to say that an event/reading's association to the device, profile, and device resources is, after the time of creation, not guaranteed to necessarily be associated to one another again after the initial creation.  This seems problematic when the event/reading gets to the north side and someone needs to get more metadata information but the associations are ephemeral.